### PR TITLE
Updated react-router 5.0->5.3 

### PR DESCRIPTION
**Pull request by rianfowler**
#### What this PR does / why we need it:
- upgrades kots from react-router 5.0->5.1 

#### Which issue(s) this PR fixes:
https://app.shortcut.com/replicated/story/59398/update-kots-web-react-router-5-0-5-1

#### Does this PR introduce a user-facing change?

```release-note
- upgrade kots/web react-router 5.0 -> 5.3
```

#### Does this PR require documentation?
none

### CODE REVIEW

File: web/src/utilities/react-router-utilities.tsx [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+import React from "react";
+import {
+  match,
+  RouteComponentProps,
+  useHistory,
+  useLocation,
+  useRouteMatch,
+} from "react-router";
+
+// @ts-ignore
```
Comment (rianfowler) [inline web/src/utilities/react-router-utilities.tsx:11]: 🤷🏽

File: web/src/utilities/react-router-utilities.tsx [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+const withRouter = <P extends object>(
+  WrappedComponent: React.ComponentType<
+    RouteComponentProps | { wrappedMatch: match } | P
+  >
+) => {
+  return class extends React.Component<P> {
+    render() {
+      return (
+        <RouterWrapper
+          // @ts-ignore
```
Comment (rianfowler) [inline web/src/utilities/react-router-utilities.tsx:32]: I'm not good enough TS to make this work.

File: web/src/utilities/react-router-utilities.tsx [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+              />
+            );
+          }}
+        />
+      );
+    }
+  };
+};
+
+export type withRouterType = RouteComponentProps & {
```
Comment (rianfowler) [inline web/src/utilities/react-router-utilities.tsx:50]: `AppVersionHistory` is the only thing that needs this.

File: web/src/components/PreflightResultPage.jsx [hunk 4]
```diff
@@ -1,7 +1,7 @@
 import React, { Component } from "react";
 import { Link } from "react-router-dom";
 import { KotsPageTitle } from "@components/Head";
-import { withRouter } from "react-router-dom";
+import { withRouter } from "@src/utilities/react-router-utilities";
```
Comment (rianfowler) [inline web/src/components/PreflightResultPage.jsx:4]: most changes are just swapping in the `withRouter` from our utility .

File: web/src/components/apps/AppDetailPage.tsx [hunk 5]
```diff
@@ -559,25 +559,23 @@ function AppDetailPage(props: Props) {
...
                 />
-                <Redirect
-                  exact
-                  from="/app/:slug/snapshots/:id/restore"
-                  to="/snapshots/partial/:slug/:id/restore"
+                <Route
+                  path="/app/:slug/snapshots/:id/restore"
+                  render={() => (
+                    <Redirect to="/snapshots/partial/:slug/:id/restore" />
+                  )}
```
Comment (rianfowler) [inline web/src/components/apps/AppDetailPage.tsx:578]: Following the insruction laid out here: https://reactrouter.com/en/v6.3.0/upgrading/v5#remove-redirects-inside-switch
Screen Shot 2022-11-16 at 09 26 22

File: web/src/components/apps/AppVersionHistory.tsx [hunk 6]
```diff
@@ -240,7 +238,7 @@ class AppVersionHistory extends Component<Props, State> {
     this.state.appUpdateChecker.start(this.getAppUpdateStatus, 1000);
 
     const url = window.location.pathname;
-    const { params } = this.props.match;
+    const { params } = this.props.wrappedMatch;
```
Comment (rianfowler) [inline web/src/components/apps/AppVersionHistory.tsx:241]: I could not figure out what was going on with the `match` for this component but it's getting modified somewhere.

File: web/src/utilities/react-router-utilities.tsx [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+const RouterWrapper = ({ children }) => {
+  const history = useHistory();
+  const location = useLocation();
+  const wrappedMatch = useRouteMatch<KotsParams>();
+
+  return children({ history, location, wrappedMatch });
+};
+
+/**
+ * @deprecated The method should not be used on new components. New components should use the hooks directly.
```
Comment (rianfowler) [inline web/src/utilities/react-router-utilities.tsx:21]: this causes the `withRouter` to show up crossed out with this comment in vscode.

**Review by miaawong — APPROVED** [review]
awesome!

**Pull request merged**